### PR TITLE
tshttpproxy: degrade gracefully when WPAD is unavailable

### PR DESCRIPTION
**Pull request by deansheather**
## Problem

On Windows, tailscaled invokes `WinHttpGetProxyForUrl` on every outbound HTTP request to discover a proxy via WPAD (DHCP option 252 / DNS A `wpad.<domain>`). On networks where WPAD is **not** configured, this call blocks for up to 5s and returns `ERROR_WINHTTP_AUTODETECTION_FAILED` or `ERROR_WINHTTP_UNABLE_TO_DOWNLOAD_SCRIPT`.

The existing 10s negative-cache window was repeatedly torn down by `InvalidateCache()` on every netmon link change, so control / DERP / log uploads spent most of their time waiting on dead WPAD probes on hosts without WPAD.

Goal: **use WPAD if it's available; fall back to direct (or env-var/static proxy) if it's not.** No WPAD requirement.

References upstream: tailscale/tailscale#17055, tailscale/tailscale#10215.

## Changes

All in `net/tshttpproxy/tshttpproxy_windows.go`:

1. **Backoff bumped from 10s → 5 minutes** for `ERROR_WINHTTP_AUTODETECTION_FAILED` and `ERROR_WINHTTP_UNABLE_TO_DOWNLOAD_SCRIPT`. `InvalidateCache()` still resets it on real link changes, so legitimate WPAD setups keep working — we just stop re-blocking on a 5s probe for every individual HTTP request.
2. **Context-timeout no longer returns stale proxy.** Previously the 5s deadline path returned `cachedProxy.val` from a prior network — wrong answer when roaming onto a network without WPAD. Now returns `(nil, nil)` with a 30s backoff.
3. **Unknown WinHTTP errors no longer propagate.** Mapped to `(nil, nil)` + 5min backoff. The package-level `ProxyFromEnvironment` already swallows `sysProxyFromEnv` errors via shadowed-err in its wrapper, but making the contract explicit at the platform layer documents intent and prevents regressions.

WPAD continues to work unchanged on networks where it is configured; this PR only changes how we degrade when it is not.

## Test

Adds a cross-platform regression test (`TestProxyFromEnvironment_sysProxyError`) that injects a failing `sysProxyFromEnv` hook and asserts `ProxyFromEnvironment` returns `(nil, nil)`, respects an active `noProxyUntil` backoff, and re-probes after `InvalidateCache()`.

## Validation

- `go test ./net/tshttpproxy/...` ✅
- `go build ./...` (Linux) ✅
- `GOOS=windows go build ./cmd/tailscaled/... ./cmd/tailscale/...` ✅
- `GOOS=windows go vet ./net/tshttpproxy/...` ✅

## Scope intentionally not included

- `WinHttpGetIEProxyConfigForCurrentUser` fast-path that would skip WinHTTP entirely when WPAD is administratively disabled in IE settings / group policy.
- Honoring static IE proxy (`lpszProxy`) when only "Use a proxy server" is configured.

Both are pure optimizations and require new syscall bindings + `zsyscall_windows.go` regeneration. The user-visible goal (don't require WPAD) is achieved by this PR; those are candidates for a follow-up if telemetry shows it's worth the diff.

## Behavior change to call out

If a customer today is failing fast due to an unmapped WinHTTP error, they will now silently dial direct instead. The error is still logged (rate-limited) and `winhttp_proxy_err_other` still increments, so it's observable.

### CODE REVIEW

File: net/tshttpproxy/tshttpproxy_windows.go [hunk 1]
```diff
@@ -86,13 +105,27 @@ func proxyFromWinHTTPOrCache(req *http.Request) (*url.URL, error) {
...
+		// tailscale/tailscale#10215.
 		const (
 			ERROR_WINHTTP_AUTODETECTION_FAILED      = 12180
 			ERROR_WINHTTP_UNABLE_TO_DOWNLOAD_SCRIPT = 12167
 		)
 		if err == syscall.Errno(ERROR_WINHTTP_AUTODETECTION_FAILED) {
 			metricErrDetectionFailed.Add(1)
-			setNoProxyUntil(10 * time.Second)
+			setNoProxyUntil(wpadAutodetectFailedBackoff)
 			return nil, nil
```
Comment (deansheather) [inline net/tshttpproxy/tshttpproxy_windows.go:148]: **P1** Stale in-flight WPAD probe can stomp a fresh `InvalidateCache()` and lock proxy discovery out for the full 5-minute backoff *(Concurrency Reviewer P1)*

> Interleaving:
> 1. T=0: Request R1 calls `proxyFromWinHTTPOrCache`. Goroutine G1 starts the blocking `WinHttpGetProxyForUrl` (5s ctx).
> 2. T=2s: A netmon link change fires. `InvalidateCache()` takes `mu`, sets `noProxyUntil = time.Time{}`. The user is now on a network where WPAD might actually work.
> 3. T=4s: G1 finally returns with `ERROR_WINHTTP_AUTODETECTION_FAILED` from the *previous* network. The select takes the result branch and calls `setNoProxyUntil(wpadAutodetectFailedBackoff)` = 5 minutes. `noProxyUntil` is now `now+5min`, derived from a probe issued before the link change.
> 4. For the next 5 minutes on the new network, every caller in `ProxyFromEnvironment` takes the `time.Now().Before(noProxyTime)` branch and returns `(nil, nil)` without ever consulting WPAD on the new network.

This race existed pre-PR but was bounded to ~10s — usually invisible. With the new 5-minute backoff, the same race becomes a user-visible "WPAD doesn't work after I roamed networks" regression that lasts 30× longer. The whole point of the PR is to make WPAD-less hosts work; the symmetric failure (host *gains* WPAD and can't use it) is the kind of thing that's tolerable at 10s and not at 5min.

Fix shape: bump a generation/epoch counter inside `InvalidateCache`, snapshot it at goroutine launch, and skip the `setNoProxyUntil` call in the result branch when the epoch has changed since launch. Equivalent: track `lastInvalidate time.Time` and make `setNoProxyUntil` a no-op when the in-flight probe started before `lastInvalidate`.

File: net/tshttpproxy/tshttpproxy_windows.go [hunk 2]
```diff
@@ -106,17 +139,24 @@ func proxyFromWinHTTPOrCache(req *http.Request) (*url.URL, error) {
...
-		defer cachedProxy.Unlock()
-		proxyErrorf("tshttpproxy: winhttp: GetProxyForURL(%q): timeout; using cached proxy %v", urlStr, cachedProxy.val)
-		return cachedProxy.val, nil
+		// Coder fork: previously returned cachedProxy.val, which
+		// stranded the client on a stale proxy from a previous
+		// network. Return no-proxy and apply a short backoff so
+		// in-flight WinHTTP probes don't pile up on every request
+		// while WPAD is slow.
+		setNoProxyUntil(wpadTimeoutBackoff)
+		proxyErrorf("tshttpproxy: winhttp: GetProxyForURL(%q): timeout; falling back to direct", urlStr)
```
Comment (deansheather) [inline net/tshttpproxy/tshttpproxy_windows.go:186]: **P2** Timeout path now silently dials direct for 30s on networks that genuinely need a proxy *(Edge Case Analyst P2, Contract Auditor P2)*

> Pre-PR, on `ctx.Done()` the function returned `cachedProxy.val` (the last successful WPAD result). Post-PR it returns `(nil, nil)` and sets a 30s backoff. For users on networks where WPAD genuinely returns a proxy but `WinHttpGetProxyForURL` occasionally exceeds the 5s budget (slow PAC fetch, contended thread pool, AV hooking the call), every timed-out request now dials direct for 30 seconds. That can (a) fail outright if egress is firewalled to proxy-only, and (b) leak traffic that was meant to traverse the corporate proxy (TLS to external hosts) past DLP/logging.

> *Contract Auditor adds:* link changes already call `InvalidateCache()` and clear `cachedProxy` indirectly via the next successful lookup. On the same network, the old behavior was the safer choice.

The commit message frames the old behavior as "stranded the client on a stale proxy from a previous network" — that's the failure mode after a link change, not on the same network. The symmetric risk (silent direct-dial for proxy-required users) is the worse of the two and isn't called out.

A recency- and epoch-gated fallback would resolve both: on timeout, return `cachedProxy.val` only when it was set during the current network epoch (see the P1 epoch suggestion) and within a small recency window; otherwise fall through to direct.

File: net/tshttpproxy/tshttpproxy_windows.go [hunk 3]
```diff
@@ -106,17 +139,24 @@ func proxyFromWinHTTPOrCache(req *http.Request) (*url.URL, error) {
...
+		return nil, nil
```
Comment (deansheather) [inline net/tshttpproxy/tshttpproxy_windows.go:187]: **P2** `cachedProxy` is now write-only state — the name promises a cache but nothing reads it *(Contract Auditor P2, Edge Case Analyst P3, Go Architect P3, Concurrency Reviewer Obs)*

> After this PR, `cachedProxy.val` is written on every success but never read anywhere except the log-diff at line 100. The timeout path used to be its sole reader and now ignores it. A reader seeing `cachedProxy` will assume it's consulted on lookup failure, and a future change "wiring up the cache" will accidentally reintroduce stale-proxy behavior the author of this PR is trying to remove.

Resolution depends on the P2 above: if the timeout path keeps a recency/epoch-gated fallback, `cachedProxy` stays alive and earns its name. If not, delete the struct (and the success-path Lock/diff logging that depends on it) or rename to `lastObservedProxy` so the role matches the name. Leaving it as written is a trap for the next maintainer.

File: net/tshttpproxy/tshttpproxy_windows.go [hunk 4]
```diff
@@ -86,13 +105,27 @@ func proxyFromWinHTTPOrCache(req *http.Request) (*url.URL, error) {
...
+		// ProxyFromEnvironment already discards sysProxyFromEnv errors,
+		// so propagating one here was effectively dead code — but
+		// returning nil explicitly documents the contract and keeps us
+		// honest if that wrapper ever changes. The real bug being fixed
+		// is the backoff durations being far too short: with the
+		// previous 10s, a host with no WPAD server (DHCP option 252
+		// unset, no wpad.<domain> A record) would re-issue a 5s blocking
+		// WinHttpGetProxyForUrl on nearly every outbound HTTP request,
+		// because netmon link changes call InvalidateCache() and reset
+		// the shield constantly. See tailscale/tailscale#17055 and
```
Comment (deansheather) [inline net/tshttpproxy/tshttpproxy_windows.go]: **P2** The new comment claims `ProxyFromEnvironment` already discards `sysProxyFromEnv` errors, but that discard is variable shadowing in a different file, not an explicit contract *(Go Architect P2, Edge Case Analyst Obs, Test Auditor Obs, Concurrency Reviewer Obs)*

> `tshttpproxy.go:144` returns `err` from the env-var `localProxyFunc`, not from `sysProxyFromEnv`. The sys error is discarded only because line 138-141 reassigns `err` to a shadowed local. Today that means a sys error is dropped, yes — but the comment in `tshttpproxy_windows.go` reads as if the wrapper has a stable contract. It does not: a one-line refactor of `ProxyFromEnvironment` (e.g., un-shadowing `err`) would silently re-propagate WinHTTP errors and undo this PR's intent without touching the Windows file.

The load-bearing piece of this PR's behavior is split across two files and one of them depends on a coincidence. Worth fixing in the same PR: drop the shadowed `err` in `tshttpproxy.go:138` (`u, _ := sysProxyFromEnv(req)`) or restructure so the sys-error path explicitly returns `nil, nil`. Then the Windows file's comment can simply say "honor the package contract" and the contract actually exists. Otherwise add an end-to-end regression test that catches a future un-shadowing.

File: net/tshttpproxy/tshttpproxy_test.go [hunk 5]
```diff
@@ -205,3 +206,78 @@ func TestSetSelfProxy(t *testing.T) {
...
 }
+
+// TestProxyFromEnvironment_sysProxyError verifies that when the
+// platform-specific proxy lookup (e.g. Windows WPAD via WinHTTP) returns
+// an error, ProxyFromEnvironment surfaces (nil, nil) to the caller so
+// http.Transport falls back to a direct connection rather than failing
+// the request. This is the cross-platform glue side of the
+// coder/tailscale fix for tailscale/tailscale#17055 / #10215.
+func TestProxyFromEnvironment_sysProxyError(t *testing.T) {
+	// Make sure no env-var proxy interferes with the fallthrough.
```
Comment (deansheather) [inline net/tshttpproxy/tshttpproxy_test.go:217]: **P1** New test exercises pre-existing wrapper behavior, not the substantive Windows changes *(Test Auditor P1)*

> The Windows file changes four observable behaviors:
> 1. `setNoProxyUntil` constants jump from 10s to 5m / 30s / 5m.
> 2. The timeout branch no longer returns `cachedProxy.val`; it now returns `(nil, nil)` after a 30s backoff.
> 3. The "unmapped WinHTTP error" branch now returns `(nil, nil)` with a 5m backoff instead of `(nil, err)`.
> 4. The autodetect-fail and download-fail branches use the new long backoffs.
>
> The added cross-platform test exercises *none* of these. Apply the "delete the feature, does the test still pass?" check: revert `tshttpproxy_windows.go` entirely. The new test still passes, because nothing in it touches Windows code or the new constants.

The deep-review just produced a P2 regression in the timeout path (silent direct-dial for proxy-required users) — i.e. precisely the change with no test coverage. Factoring `proxyFromWinHTTPOrCache` to take an injectable `getProxyForURL` (and clock) would let table-driven tests cover autodetect-fail / download-fail / unknown-error / timeout against both the returned `(url, err)` and the `noProxyUntil` advance. Without that the riskiest change in the PR ships untested.

At minimum: an end-to-end Windows test asserting that on `ctx.Done()` we return `(nil, nil)` and not the previously-cached value would have caught the timeout regression directly.

File: net/tshttpproxy/tshttpproxy_windows.go [hunk 6]
```diff
@@ -106,17 +139,24 @@ func proxyFromWinHTTPOrCache(req *http.Request) (*url.URL, error) {
...
-			setNoProxyUntil(10 * time.Second)
+			setNoProxyUntil(wpadDownloadFailedBackoff)
 			return nil, nil
 		}
 		metricErrOther.Add(1)
-		return nil, err
+		// Coder fork: do not propagate unmapped WinHTTP errors. Treat
+		// them as "no proxy discovered" so the caller dials direct.
+		setNoProxyUntil(wpadUnknownErrorBackoff)
+		return nil, nil
```
Comment (deansheather) [inline net/tshttpproxy/tshttpproxy_windows.go:168]: **P3** 5-minute shield on unmapped-error path overreacts to single-shot transient errors *(Contract Auditor P3)*

> Previously `metricErrOther` returned `nil, err` with no backoff, so a one-off transient WinHTTP error caused a single direct-dial fallthrough at the package level and the next request retried WPAD immediately. Now the same one-off error suppresses WPAD for 5 full minutes. Combined with `InvalidateCache()` only firing on link changes, a host that hits one weird WinHTTP error during a flaky moment loses proxy autodetection for 5 minutes even though the network is fine.

Not blocking — pick a number, document the tradeoff. 1 minute matches the spirit of the change (avoid hot-looping on persistent failures) without overreacting to single-shot ones. If the P1 epoch fix lands, this becomes mostly moot.

File: net/tshttpproxy/tshttpproxy_windows.go [hunk 7]
```diff
@@ -52,6 +52,25 @@ var (
...
+// record) would re-issue a 5-second blocking WinHttpGetProxyForUrl call
+// for nearly every outbound HTTP request. Bumping the negative-cache
+// duration to several minutes lets normal traffic flow while still
+// re-trying WPAD periodically and on every link change (the latter via
+// InvalidateCache, which zeros noProxyUntil).
+//
+// See tailscale/tailscale#17055 and tailscale/tailscale#10215.
+const (
+	wpadAutodetectFailedBackoff = 5 * time.Minute
+	wpadDownloadFailedBackoff   = 5 * time.Minute
```
Comment (deansheather) [inline net/tshttpproxy/tshttpproxy_windows.go]: **Nit** Inconsistent constant naming and `ERROR_INVALID_PARAMETER` left as a hard-coded outlier *(Style Reviewer, Edge Case Analyst P3, Go Architect P2)*

> `wpadAutodetectFailedBackoff`, `wpadDownloadFailedBackoff`, `wpadUnknownErrorBackoff` all describe a failure category, but `wpadTimeoutBackoff` drops the `Failed`/`Error` suffix. When scanning the const block, the odd one out reads as if a timeout is somehow not a failure.

> *Edge Case Analyst adds:* the `ERROR_INVALID_PARAMETER` path still hard-codes `time.Hour` inline. The new constant block invites the question: either fold it in (`wpadInvalidParameterBackoff = time.Hour` with the Win 8.1 comment) or note in the comment why it's intentionally an outlier.

Minor; renaming `wpadTimeoutBackoff` → `wpadTimeoutFailedBackoff` (or dropping the suffix from the others) and naming the 1h constant would make the policy reviewable in one place.

File: net/tshttpproxy/tshttpproxy_windows.go (hunk 4 above)
Comment (deansheather) [inline net/tshttpproxy/tshttpproxy_windows.go]: **Nit** `Coder fork:` block comment in the function body largely restates the package-level const-block rationale *(Style Reviewer)*

> Both comments explain the same WPAD/InvalidateCache/10s-backoff history and cite the same two issues. Once a reader has seen the const-block rationale, this second copy is noise on the hot path of the function.

Trim to a one-liner pointing at the constants: `// See backoff constants above; all error paths return (nil, nil) so the caller dials direct.` The aside `keeps us honest if that wrapper ever changes` is speculation that doesn't belong in a code comment — the un-shadowing fix in the P2 above replaces this need entirely.

File: net/tshttpproxy/tshttpproxy_test.go [hunk 8]
```diff
@@ -205,3 +206,78 @@ func TestSetSelfProxy(t *testing.T) {
...
+		noProxyUntil = time.Time{}
+		mu.Unlock()
+	})
+
+	prev := sysProxyFromEnv
+	t.Cleanup(func() { sysProxyFromEnv = prev })
+
+	var calls int
+	wantErr := errors.New("synthetic WPAD failure")
+	sysProxyFromEnv = func(*http.Request) (*url.URL, error) {
```
Comment (deansheather) [inline net/tshttpproxy/tshttpproxy_test.go:244]: **Nit** `wantErr` is misleading — the test never asserts on it *(Style Reviewer)*

> `wantErr` is the conventional Go testing name for an expected-error value the test checks against (`if err != wantErr`). Here the whole point of the test is that the synthetic error is *swallowed*, so a name like `syntheticErr` or `fakeErr` matches intent and avoids misleading future readers who grep for `wantErr` patterns.

**Review by deansheather — COMMENTED** [review]
Solid diagnosis of the WPAD-on-non-WPAD-network problem, and the longer backoffs + named constants land cleanly. The deep-review found a couple of structural concerns worth a second look before this merges, plus some test coverage gaps in the exact paths the PR changes.

This review contains findings that may need attention before merge. 1 P1, 4 P2, 2 P3, 4 nits across 9 inline comments.

File: net/tshttpproxy/tshttpproxy_windows.go (hunk 1 above)
Comment (deansheather) [reply net/tshttpproxy/tshttpproxy_windows.go:148]: Fixed in cf8ada465. Added an `epoch` counter to `tshttpproxy.go` (incremented by `InvalidateCache`), exposed via `CurrentEpoch()`, and snapshotted at the start of every WPAD probe in `proxyFromWinHTTPOrCache`. The error/timeout branches now go through a `setBackoff` closure that no-ops when `CurrentEpoch() != startEpoch`, so a stale probe can no longer stomp `noProxyUntil` after a link change. Added `TestProxyFromWinHTTPOrCache_EpochStompProtection` as the explicit regression test.

File: net/tshttpproxy/tshttpproxy_windows.go (hunk 2 above)
Comment (deansheather) [reply net/tshttpproxy/tshttpproxy_windows.go:186]: Fixed in cf8ada465. The timeout branch now reads `cachedProxy` and returns the cached value when it was stamped under the current epoch (i.e. discovered on the same network we're still on). Stale-epoch or empty cache still falls through to direct with the short backoff, so the original 'stranded on a stale proxy after roaming' concern is also handled. Covered by `TestProxyFromWinHTTPOrCache_TimeoutFreshCached` and `TestProxyFromWinHTTPOrCache_TimeoutStaleCached`.

File: net/tshttpproxy/tshttpproxy_windows.go (hunk 3 above)
Comment (deansheather) [reply net/tshttpproxy/tshttpproxy_windows.go:187]: Resolved by re-introducing a real reader for `cachedProxy` in the timeout branch (epoch-gated). Also added a struct comment so the next reader knows what role it plays. Covered by the same two tests as above.

File: net/tshttpproxy/tshttpproxy_windows.go (hunk 4 above)
Comment (deansheather) [reply net/tshttpproxy/tshttpproxy_windows.go]: Fixed in cf8ada465. `ProxyFromEnvironment` now uses `if su, serr := sysProxyFromEnv(req); su != nil && serr == nil` instead of relying on shadowing, with a doc comment that explicitly states the contract: sys-proxy errors are dropped, env-proxy errors propagate. Trimmed the now-redundant block comment on the Windows side.

File: net/tshttpproxy/tshttpproxy_test.go (hunk 5 above)
Comment (deansheather) [reply net/tshttpproxy/tshttpproxy_test.go:217]: Fixed in cf8ada465. Added `tshttpproxy_windows_test.go` with seven tests covering: success (with epoch stamp), autodetect-failed, unmapped error, timeout-no-cache, timeout-fresh-cache, timeout-stale-cache, and the epoch-stomp regression. Made it possible by extracting the WinHTTP call site into a `winHTTPLookup` package var that tests can swap; the production path still calls `proxyFromWinHTTP` unchanged.

File: net/tshttpproxy/tshttpproxy_windows.go (hunk 6 above)
Comment (deansheather) [reply net/tshttpproxy/tshttpproxy_windows.go:168]: Adopted: dropped to 1 minute. Renamed to `wpadUnknownErrorFailedBackoff` for consistency with the rest of the constants block.

File: net/tshttpproxy/tshttpproxy_windows.go (hunk 7 above)
Comment (deansheather) [reply net/tshttpproxy/tshttpproxy_windows.go]: Fixed in cf8ada465. Renamed `wpadTimeoutBackoff` → `wpadTimeoutFailedBackoff` for consistency, and folded the `ERROR_INVALID_PARAMETER` 1-hour value into the named-constant block as `wpadInvalidParameterFailedBackoff` with a comment explaining the Win 8.1 origin.

File: net/tshttpproxy/tshttpproxy_windows.go (hunk 4 above)
Comment (deansheather) [reply net/tshttpproxy/tshttpproxy_windows.go]: Trimmed in cf8ada465. The const block now carries the rationale once; the function body just notes that all error paths return `(nil, nil)` per the package-level `ProxyFromEnvironment` contract. Removed the speculative 'keeps us honest' aside — the explicit `_` discard in the wrapper makes the contract real.

File: net/tshttpproxy/tshttpproxy_test.go (hunk 8 above)
Comment (deansheather) [reply net/tshttpproxy/tshttpproxy_test.go:244]: Renamed to `syntheticErr` in cf8ada465.
